ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Queries.scala

```diff
@@ -354,18 +358,22 @@ object Queries {
     oaa.copy(tail = oaa.tail.flatMap(Vector(a, _)))
 
   // Like groupBy but split into n maps where n is the longest list under groupBy.
-  // Invariant: every element of the result is non-empty
-  private[dbbackend] def uniqueSets[A, B](iter: Iterable[(A, B)]): Seq[Map[A, B]] =
-    unfold(iter.groupBy(_._1).transform((_, i) => i.toList): Map[A, List[(_, B)]]) { m =>
-      // invariant: every value of m is non-empty
-      m.nonEmpty option {
-        val hd = m transform { (_, abs) =>
-          val (_, b) +: _ = abs
-          b
+  private[dbbackend] def uniqueSets[A, B](iter: Iterable[(A, B)]): Seq[NonEmpty[Map[A, B]]] =
+    unfold(
+      iter
+        .groupBy1(_._1)
+        .transform((_, i) => i.toList): Map[A, NonEmpty[List[(_, B)]]]
+    ) {
+      case NonEmpty(m) =>
+        Some {
+          val hd = m transform { (_, abs) =>
+            val (_, b) +-: _ = abs
+            b
+          }
+          val tl = m collect { case (a, _ +-: NonEmpty(tl)) => (a, tl) }
+          (hd, tl)
         }
-        val tl = m collect { case (a, _ +: (tl @ (_ +: _))) => (a, tl) }
```

Comment:
An example bug for @SamirTalwar-DA: imagine that I had forgotten to filter out the empty `tl`s here; they would then leak into the next iteration and crash where the assumption is made on old line 363. Embedding the `NonEmpty` type into the loop state requires me to do this check and filtering instead of just calling `tail`, so the assumption on new line 370 can be made in perfect safety.

Reply:
Nice, I like it.